rep-2004.rst

```diff
@@ -0,0 +1,676 @@
+      #. *Must register with a centralized list of 'Level N' packages, if one exists, to allow for peer review of the claim*
+
+#. **Testing:**
+
+   #. Must have system tests which cover all items in the "feature" documentation
+   #. Must have system, integration, and/or unit tests which cover all of the public API
+   #. Code Coverage:
+
+      #. *Must have code coverage tracking for the package*
+      #. *Must have and enforce a code coverage policy for new changes*
+
+   #. Performance:
+
+      #. *Must have performance tests (exceptions allowed if they don't make sense to have)*
+      #. *Must have a performance regression policy (i.e. blocking either changes or releases on unexpected performance regressions)*
+
+   #. Linters and Static Analysis
+
+      #. *Must have a code style and enforce it*
+      #. *Must use static analysis tools where applicable*
+
+#. **Dependencies:**
+
+   #. Must not have direct runtime "ROS" dependencies which are not at the same level as the package in question ('Level N'), but...
+   #. May have optional direct runtime "ROS" dependencies which are not 'Level N', e.g. tracing or debugging features that can be disabled
+   #. Must have justification for why each direct runtime "non-ROS" dependency is equivalent to a 'Level N' package in terms of quality
+
+#. **Platform Support:**
+
+   #. Must support all tier 1 platforms for ROS 2, as defined in `REP-2000 <https://www.ros.org/reps/rep-2000.html#support-tiers>`_
+
+If the above points are satisfied then a package can be considered 'Level 1'.
+Refer to the detailed description of the requirements below the chart for more information.
+
+Quality Level Comparison Chart
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The chart below compares Quality Levels 1 through 5 relative to the 'Level 1' requirements' numbering scheme above.
+
+✓ = required
+
+● = recommended
+
+.. list-table:: Quality Levels
+    :widths: 5 10 10 10 10 10
+    :header-rows: 1
+    :stub-columns: 1
+    :align: left
+
+
+    * -
+      - Level 1
+      - Level 2
+      - Level 3
+      - Level 4
+      - Level 5
+    * - 1.i
+      - ✓
+      - ✓
+      - ✓
+      - ●
+      -
+    * - 1.ii
+      - ✓
+      - ✓
+      - ✓
+      -
+      -
+    * - 1.iii
+      - ✓
+      - ✓
+      - ●
+      -
+      -
+    * - 1.iv
+      - ✓
+      - ✓
+      - ✓
+      -
+      -
+    * - 1.v
+      - ✓
+      - ✓
+      - ✓
+      -
+      -
+    * - 1.vi
+      - ✓
+      - ✓
+      - ●
+      -
+      -
+    * - 2.i
+      - ✓
+      - ✓
+      - ✓
+      - ●
+      -
+    * - 2.ii
+      - ✓
+      - ✓
+      -
+      -
+      -
+    * - 2.iii
+      - ✓
+      -
+      -
+      -
+      -
+    * - 2.iv
+      - ✓
+      - ✓
+      - ✓
+      -
+      -
+    * - 2.v
+      - ✓
+      -
+      -
+      -
+      -
+    * - 3.i
+      - ✓
+      - ✓
+      -
+      -
+      -
+    * - 3.ii
+      - ✓
+      -
+      -
+      -
+      -
+    * - 3.iii
+      - ✓
+      - ✓
+      - ✓
+      - ✓
+      -
```

Comment:
```suggestion
      - ✓
```
The text below suggests that even level 5 packages should have licenses.

Reply:
Yes you're right, that text is a bit opinionated, and is meant to be a stern, but it's still a recommendation in my eyes. Software lacking a license would by default fall into level 5 too. But either way it shouldn't be blank here I think.

@maryaB-osr do you think this should be a check or a circle (indicating a recommendation)?

Reply:
> Packages in this category cannot even meet the simple 'Level 4' requirements, and for that reason should not be used.
All packages should have at least a declared license or set of licenses and should include copyright statements in each file.

I understood Level 5's description to mean that level 5 packages don't meet level 4 qualifications, so they should not be used _because_ packages should at least have license and copyright statements in order to be usable, and they don't even have that.

If you're recommending that even level 5 packages have license and copyright statements, I think the level 5 description needs to be reworded to express that. And then there won't be a level to describe packages that don't have that documentation. Additionally, will it still be fair to say level 5 packages "should not be used" even though they may or may not have a license and copyright declaration?

Reply:
This is what the current doc says about QL 5:

> Quality Level 5
> ^^^^^^^^^^^^^^^
>
> Packages in this category cannot even meet the simple 'Level 4' requirements, and for that reason should not be used.
> All packages should have at least a declared license or set of licenses and should include copyright statements in each file.

So it already says it should not be used.

The last sentence is just an opinionated thing essentially equivalent to:

> If it doesn't have a license and/or copyright statement then it isn't software.

Which is obviously not true, but again it's stern and opinionated.

We could reword the last sentence to be:

> Quality level 5 has no requirements, but we still strongly recommend all packages have a license and copyright statement, ideally in each file.

Or something similar?

Reply:
That makes sense. I still find it odd that it goes from saying packages in this category should not be used to recommending requirements for it. It seems like a throwaway category.

I would suggest just getting rid of the first sentence because it seems to contradict the recommendation, both by saying it doesn't meet level 4 (level 4's only requirement is a license and copyright statement) and by saying it shouldn't be used.

> Quality level 5 has no requirements, but we still strongly recommend all packages have a license and copyright statement, ideally in each file.
A package without a license and copyright statement should not be used.